src/extension/panel.ts

```diff
@@ -0,0 +1,162 @@
+                    ReactDOM.render(
+                        React.createElement(Index, { store: new Store(${JSON.stringify(state)}) }),
+                        document.getElementById('root'),
+                    )
+                </script>
+            </body>
+            </html>`;
+
+        webview.onDidReceiveMessage(async message => {
+            if (!message) return;
+            switch (message.command) {
+            case 'open': {
+                const uris = await window.showOpenDialog({
+                    defaultUri: workspace.workspaceFolders?.[0]?.uri,
+                    filters: { 'SARIF files': ['sarif', 'json'] },
+                });
+                if (!uris) return;
+                store.logs.push(...await loadLogs(uris));
+                break;
+            }
+            case 'removeLog': {
+                store.logs.removeWhere(log => log._uri === message.uri);
+                break;
+            }
+            case 'select': {
+                const {logUri, uri, region} = message as { logUri: string, uri: string, region: _Region};
+                const validatedUri = await basing.translateArtifactToLocal(uri);
+                if (!validatedUri) return;
+                await this.selectLocal(logUri, validatedUri, region);
+                break;
+            }
+            case 'selectLog': {
+                const [logUri, runIndex, resultIndex] = message.id as ResultId;
+                const log = store.logs.find(log => log._uri === logUri);
+                const result = store.logs.find(log => log._uri === logUri)?.runs[runIndex]?.results?.[resultIndex];
+                if (!log || !result) return;
+                await this.selectLocal(logUri, log._uriUpgraded ?? log._uri, result._logRegion);
+                break;
+            }
+            case 'setState': {
+                const oldState = Store.globalState.get('view', defaultState);
+                const {state} = message;
+                await Store.globalState.update('view', Object.assign(oldState, JSON.parse(state)));
+                break;
+            }
+            default:
+                throw new Error(`Unhandled command: ${message.command}`,);
+            }
+        }, undefined, context.subscriptions);
+
+        await this.spliceLogs([], store.logs);
+    }
+
+    public async selectLocal(logUri: string, localUri: string, region: _Region | undefined) {
+        // Keep/pin active Log as needed
+        for (const editor of window.visibleTextEditors.slice()) {
+            if (editor.document.uri.toString() !== logUri) continue;
+            await window.showTextDocument(editor.document, editor.viewColumn);
+            await commands.executeCommand('workbench.action.keepEditor');
+        }
+
+        const doc = await workspace.openTextDocument(Uri.parse(localUri));
+        const editor = await window.showTextDocument(doc, ViewColumn.One, true);
+
+        if (region === undefined) return;
```

Comment:
>region === undefined [](start = 12, length = 20)

Some places you're doing (!o) and others you are doing (o === undefined), would be nice to be consistent please.

Reply:
My pattern is to always use the short form, unless I can't. In this case I can't because region can be a number and 0 is a valid value.

Anecdotally the short form works 90%+ of the time and I like how it optimizes for the common case. That said, let's talk about this in the next meeting.

---
In reply to: [441653931](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441653931) [](ancestors = 441653931)